### PR TITLE
Add chainId validation to the network settings page

### PR DESCRIPTION
**Pull request by spylogsster**
Resolves https://github.com/brave/brave-browser/issues/20605

- Add chainId validation to the network settings page
image

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

- Try add a new network from the settings
- check it will alert for wrong urls and ids

### CODE REVIEW

File: browser/ui/webui/settings/brave_wallet_handler.h [hunk 1]
```diff
@@ -29,8 +39,24 @@ class BraveWalletHandler : public settings::SettingsPageUIHandler {
...
+                          const std::string& chain_id,
+                          brave_wallet::mojom::ProviderError error,
+                          const std::string& error_message);
+  brave_wallet::JsonRpcService* GetJsonRpcService();
+  void SetJsonRpcServiceForTesting(
+      brave_wallet::JsonRpcService* test_json_rpc_service) {
+    test_json_rpc_service_ = test_json_rpc_service;
+  }
+  base::OnceClosure chain_callback_for_testing_;
+  brave_wallet::JsonRpcService* test_json_rpc_service_ = nullptr;
```
Comment (yrliou) [inline browser/ui/webui/settings/brave_wallet_handler.h:58]: As mentioned in the DM, let's try to just use `JsonRpcSerive::SetAPIRequestHelperForTesting` in the test to set the url loader for intercepting requests, then we could remove this, `GetJsonRpcService`, and `SetJsonRpcServiceForTesting` all together.

File: browser/ui/webui/settings/brave_wallet_handler.cc [hunk 2]
```diff
@@ -135,12 +120,72 @@ void BraveWalletHandler::GetCustomNetworksList(
...
+    *error_message = l10n_util::GetStringUTF8(
+        IDS_SETTINGS_WALLET_NETWORKS_SUMBISSION_FAILED);
+    return false;
+  }
+  PrefService* prefs = Profile::FromWebUI(web_ui())->GetPrefs();
+  if (IsChainExist(prefs, chain->chain_id)) {
+    *error_message =
+        l10n_util::GetStringUTF8(IDS_SETTINGS_WALLET_NETWORKS_EXISTS);
+    return false;
+  }
```
Comment (yrliou) [inline browser/ui/webui/settings/brave_wallet_handler.cc:145]: I think it makes more sense to move it to `JsonRpcService::AddEthereumChain`, this check should be shared across all possible callers of that API.

File: browser/ui/webui/settings/brave_wallet_handler.cc [hunk 3]
```diff
@@ -135,12 +120,72 @@ void BraveWalletHandler::GetCustomNetworksList(
   ResolveJavascriptCallback(args[0], base::Value(json_string));
 }
 
+bool BraveWalletHandler::AddEthereumChainToRPCService(
```
Comment (yrliou) [inline browser/ui/webui/settings/brave_wallet_handler.cc:123]: This naming seems a bit weird to me, I would suggest changing `absl::optional<base::Value> GetEthereumChain` you have above to absl::optional<brave_wallet::mojom::EthereumChain> GetEthereumChain`, and move the `brave_wallet::ValueToEthereumChain` part in this function into it. All these error messages should be set directly in that function too.

Comment (yrliou) [reply browser/ui/webui/settings/brave_wallet_handler.cc:123]: Also move the part for calling json_rpc_service->AddEthereumChain into `BraveWalletHandler::AddEthereumChain`, then you don't need this `AddEthereumChainToRPCService` anymore.

File: browser/ui/webui/settings/brave_wallet_handler.cc [hunk 4]
```diff
@@ -135,12 +120,72 @@ void BraveWalletHandler::GetCustomNetworksList(
...
+    base::Value javascript_callback,
+    const std::string& payload,
+    std::string* error_message) {
+  auto value_to_add = GetEthereumChain(payload, error_message);
+  if (!value_to_add) {
+    *error_message = l10n_util::GetStringUTF8(
+        IDS_SETTINGS_WALLET_NETWORKS_SUMBISSION_FAILED);
```
Comment (yrliou) [inline browser/ui/webui/settings/brave_wallet_handler.cc:130]: Isn't this unnecessary as it should have already been set in https://github.com/brave/brave-core/pull/11971/files#diff-4485cfb12a4c60f887f9e1929fc5c346c50422ee430645a68334cb360c97989fR43-R45?

File: browser/ui/webui/settings/brave_wallet_handler.h (hunk 1 above)
Comment (spylogsster) [reply browser/ui/webui/settings/brave_wallet_handler.h]: done

File: browser/ui/webui/settings/brave_wallet_handler.cc (hunk 2 above)
Comment (spylogsster) [reply browser/ui/webui/settings/brave_wallet_handler.cc]: done

File: browser/ui/webui/settings/brave_wallet_handler.cc (hunk 3 above)
Comment (spylogsster) [reply browser/ui/webui/settings/brave_wallet_handler.cc]: done

File: browser/ui/webui/settings/brave_wallet_handler.cc (hunk 4 above)
Comment (spylogsster) [reply browser/ui/webui/settings/brave_wallet_handler.cc]: done

File: browser/ui/webui/settings/brave_wallet_handler.h [hunk 5]
```diff
@@ -31,6 +41,14 @@ class BraveWalletHandler : public settings::SettingsPageUIHandler {
 
   BraveWalletHandler(const BraveWalletHandler&) = delete;
   BraveWalletHandler& operator=(const BraveWalletHandler&) = delete;
+
+  void OnAddEthereumChain(base::Value javascript_callback,
+                          const std::string& chain_id,
+                          brave_wallet::mojom::ProviderError error,
+                          const std::string& error_message);
+  brave_wallet::JsonRpcService* GetJsonRpcService();
```
Comment (yrliou) [inline browser/ui/webui/settings/brave_wallet_handler.h:49]: this is redundant now.

File: browser/ui/webui/settings/brave_wallet_handler.cc [hunk 6]
```diff
@@ -135,14 +117,39 @@ void BraveWalletHandler::GetCustomNetworksList(
...
-  PrefService* prefs = Profile::FromWebUI(web_ui())->GetPrefs();
   AllowJavascript();
   std::string error_message;
-  bool success = ::AddEthereumChain(prefs, args[1].GetString(), &error_message);
+  auto* json_rpc_service =
+      brave_wallet::JsonRpcServiceFactory::GetServiceForContext(
+          Profile::FromWebUI(web_ui()));
+
+  auto chain = GetEthereumChain(args[1].GetString(), &error_message);
+  if (chain && json_rpc_service) {
```
Comment (yrliou) [inline browser/ui/webui/settings/brave_wallet_handler.cc:143]: for json_rpc_service nullptr case, no error message is set, is it expected? I suppose we want to set the error message for any error cases?

File: browser/ui/webui/settings/brave_wallet_handler.h (hunk 5 above)
Comment (spylogsster) [reply browser/ui/webui/settings/brave_wallet_handler.h]: done

**Pull request merged**